code/game/machinery/autolathe/matterforge.dm

```diff
@@ -0,0 +1,594 @@
+			error = error_messages[err]
+		else
+			error = "Unknown error."
+
+		if(current_design && progress >= current_design.time)
+			finish_construction()
+
+	else
+		error = null
+		working = FALSE
+		next_file()
+
+	update_icon()
+	SSnano.update_uis(src)
+
+
+
+/obj/machinery/matter_nanoforge/proc/next_file()
+	current_design = null
+	progress = 0
+	if(queue.len)
+		var/hold = queue[1]
+		current_design = hold
+		print_pre()
+		working = TRUE
+		queue.Cut(1, 2) // Cut queue[1]
+	else
+		working = FALSE
+	update_icon()
+
+/obj/machinery/matter_nanoforge/proc/eject(amount)
+
+	if(!stored_material.len || stored_material[1] == 0)
+		return
+
+	if (!amount)
+		return
+
+	amount = min(amount, stored_material[MATERIAL_COMPRESSED_MATTER])
+
+	var/whole_amount = round(amount)
+	var/remainder = amount - whole_amount
+
+	if (whole_amount)
+		var/obj/item/stack/material/compressed_matter/S = new(drop_location())
+
+		//Accounting for the possibility of too much to fit in one stack
+		if (whole_amount <= S.max_amount)
+			S.amount = whole_amount
+			S.update_strings()
+			S.update_icon()
+		else
+			//There's too much, how many stacks do we need
+			var/fullstacks = round(whole_amount / S.max_amount)
+			//And how many sheets leftover for this stack
+			S.amount = whole_amount % S.max_amount
+
+			if (!S.amount)
+				qdel(S)
+
+			for(var/i = 0; i < fullstacks; i++)
+				var/obj/item/stack/material/compressed_matter/MS = new(drop_location())
+				MS.amount = MS.max_amount
+				MS.update_strings()
+				MS.update_icon()
+
+	//And if there's any remainder, we eject that as a shard
+	if (remainder)
+		new /obj/item/weapon/material/shard(drop_location(), MATERIAL_COMPRESSED_MATTER, _amount = remainder)
+
+	//The stored material gets the amount (whole+remainder) subtracted
+	stored_material[MATERIAL_COMPRESSED_MATTER] -= amount
+
+/obj/machinery/matter_nanoforge/proc/abort()
+	if(working)
+		print_post()
+	current_design = null
+	paused = TRUE
+	working = FALSE
+	update_icon()
+
+//Finishing current construction
+/obj/machinery/matter_nanoforge/proc/finish_construction()
+	fabricate_design(current_design)
+
+/obj/machinery/matter_nanoforge/proc/fabricate_design(datum/design/design)
+	consume_materials(design)
+	design.Fabricate(drop_location(), mat_efficiency, src)
+
+	working = FALSE
+	current_design = null
+	print_post()
+	next_file()
+
+/obj/machinery/matter_nanoforge/proc/consume_materials(datum/design/design)
+	for(var/material in design.materials)
+	// + 15 base cost for each material then * 2 
+		design.materials[material] += ((1 - lst[material]) * 10) * 2
+		var/material_cost = design.adjust_materials ? SANITIZE_LATHE_COST(design.materials[material]): design.materials[material]
+		stored_material[MATERIAL_COMPRESSED_MATTER] = max(0, stored_material[MATERIAL_COMPRESSED_MATTER] - material_cost)
+
+	return TRUE
+
+#undef ERR_OK
+#undef ERR_NOTFOUND
+#undef ERR_NOMATERIAL
+#undef ERR_PAUSED
+
+/obj/machinery/matter_nanoforge/proc/update_desc(var/stored_mats)
```

Comment:
Might as well do an /examine() override, and then tell the user how many stored_mats there are through that, rather than updating the desc every time you store materials.

Reply:
done